shadowsocks/dialer.go

```diff
@@ -0,0 +1,196 @@
+func (d *ssDialer) DialUDP(laddr *net.UDPAddr, raddr string) (onet.PacketConn, error) {
+	targetHost, targetPort, err := SplitHostPortNumber(raddr)
+	if err != nil {
+		return nil, err
+	}
+	targetAddr := &packetConnAddr{Host: targetHost, Port: targetPort}
+	proxyAddr := &net.UDPAddr{IP: d.proxyIP, Port: d.proxyPort}
+	pc, err := net.ListenUDP("udp", laddr)
+	if err != nil {
+		return nil, err
+	}
+	conn := packetConn{
+		PacketConn: pc, proxyAddr: proxyAddr, targetAddr: targetAddr,
+		cipher: d.cipher, buf: make([]byte, udpBufSize)}
+	return &conn, nil
+}
+
+type packetConn struct {
+	net.PacketConn
+	io.ReadWriter
+	proxyAddr  *net.UDPAddr
+	targetAddr net.Addr
+	cipher     shadowaead.Cipher
+	m          sync.Mutex
+	buf        []byte // Write lock
+}
+
+// Write encrypts `b` and writes to the connected address through the proxy.
+func (c *packetConn) Write(b []byte) (int, error) {
+	return c.WriteTo(b, c.targetAddr)
+}
+
+// WriteTo encrypts `b` and writes to `addr` through the proxy.
+func (c *packetConn) WriteTo(b []byte, addr net.Addr) (int, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	socksTargetAddr := socks.ParseAddr(addr.String())
+	if socksTargetAddr == nil {
+		return 0, errors.New("Failed to parse target address")
+	}
+	buf, err := shadowaead.Pack(c.buf, append(socksTargetAddr, b...), c.cipher)
+	if err != nil {
+		return 0, err
+	}
+	_, err = c.PacketConn.WriteTo(buf, c.proxyAddr)
+	return len(b), err
+}
+
+func (c *packetConn) Read(b []byte) (int, error) {
+	n, _, err := c.ReadFrom(b)
+	return n, err
+}
+
+// ReadFrom reads from the embedded PacketConn and decrypts into `b`.
+func (c *packetConn) ReadFrom(b []byte) (int, net.Addr, error) {
+	if len(b) < c.cipher.SaltSize() {
+		return 0, nil, shadowaead.ErrShortPacket
+	}
+	n, addr, err := c.PacketConn.ReadFrom(b)
+	if err != nil {
+		return n, nil, err
+	} else if addr.String() != c.proxyAddr.String() {
+		return n, addr, errors.New("Received a packet from an unassociated address")
+	}
+	// Avoid overlapping the destination and cipher buffers per https://golang.org/pkg/crypto/cipher/#AEAD.
+	buf, err := shadowaead.Unpack(b[c.cipher.SaltSize():], b[:n], c.cipher)
+	if err != nil {
+		return n, nil, err
+	}
+	socksSrcAddr := socks.SplitAddr(buf[:n])
+	if socksSrcAddr == nil {
+		return n, nil, errors.New("Failed to read source address")
+	}
+	srcHost, srcPort, err := SplitHostPortNumber(socksSrcAddr.String())
+	if err != nil {
+		return len(buf), nil, errors.New("Failed to parse source address")
+	}
+	srcAddr := &packetConnAddr{Host: srcHost, Port: srcPort}
+	copy(b, buf[len(socksSrcAddr):]) // Strip the SOCKS source address
+	return len(buf) - len(socksSrcAddr), srcAddr, err
+
+}
+
+// Convenience struct to hold a domain name or IP address host. Used for SOCKS addressing.
+type packetConnAddr struct {
+	net.Addr
+	Host string
+	Port int
+}
+
+func (a *packetConnAddr) String() string {
+	return net.JoinHostPort(a.Host, strconv.FormatInt(int64(a.Port), 10))
```

Comment:
Given that `Host` and `Port` are only read here, would it make sense to use `addr` instead?  This would avoid performing the string construction, on every packet, possibly saving a memory allocation.  It would also remove the need for `SplitHostPortNumber`.

Reply:
N/A, see shadowsocks/client.go